### PR TITLE
Revenuecat v3.5.1

### PR DESCRIPTION
**Pull request by ethervoid**
I've made the sync of the upstream in our fork and this PR is a rebase of our changes in the new version.

@bisho can you take a look? I'm not 100% sure the rebase is ok :)

### CODE REVIEW

File: beeline/trace.py [hunk 1]
```diff
@@ -117,18 +130,14 @@ def start_span(self, context=None, parent_id=None, is_root_span=False):
...
+        data = {
             'trace.trace_id': self._trace.id,
             'trace.parent_id': parent_span_id,
             'trace.span_id': span_id,
         }
-        if is_root_span:
-            spanType = "root"
-            if parent_span_id:
-                spanType = "subroot"
-            fields['meta.span_type'] = spanType
```
Comment (bisho) [inline beeline/trace.py:129]: This looks new

Comment (ethervoid) [reply beeline/trace.py:129]: Yep, I think that could be new

File: beeline/trace.py [hunk 2]
```diff
@@ -432,17 +453,11 @@ def inner(*args, **kwargs):
...
 
 
-# Use system random instead of default psuedorandom generator
-system_random = random.SystemRandom()
-
-
 def generate_span_id():
     """Generate span ID compatible with w3c tracing spec."""
-    format_str = "{{:0{:d}x}}".format(SPAN_ID_BYTES*2)  # pylint: disable=C0209
-    return format_str.format(system_random.getrandbits(SPAN_ID_BYTES*8))
```
Comment (bisho) [inline beeline/trace.py:442]: glad they started to use `getrandbits`, their old implementation was using crypto-level random and very expensive. Yet I tested and our version is still 4x faster

Comment (ethervoid) [reply beeline/trace.py:442]: This PR has your changes rebased on their last version

Comment (bisho) [reply beeline/trace.py:442]: Scratch that. They still use system_random that will drain the entropy pool :/

**Comment by bisho** [comment]
LG. I might merge each of the 2 improvements separately instead of a single merge, so we keep the patches clear and easy to port into future versions. But we are not likely to update this very often and we still have the old as reference, so no biggie. The port of the patches look OK to me. Let's canary to ensure works as expected.

**Comment by ethervoid** [comment]
We can test it and if works we can do an upstream PR, that would be very helpful to keep it updated without having to do it manually

File: beeline/trace.py [hunk 3]
```diff
@@ -117,18 +130,14 @@ def start_span(self, context=None, parent_id=None, is_root_span=False):
...
-        if is_root_span:
-            spanType = "root"
-            if parent_span_id:
-                spanType = "subroot"
-            fields['meta.span_type'] = spanType
-        ev.add(data=fields)
-
+        if self._trace.trace_sample_rate > 1:
+            data['trace.trace_sample_rate'] = self._trace.trace_sample_rate
+        ev.add(data=data)
```
Comment (bisho) [inline beeline/trace.py:142]: So add here?
```suggestion
        if is_root_span:
            data['meta.span_type'] = "root" if not parent_span_id else "subroot"
        ev.add(data=data)
```

**Review by bisho — APPROVED** [review]
LG

**Comment by ethervoid** [comment]
Close it because we're pointing to the branch in our requirements.in, this PR was just to review the changes

**Pull request closed without merging**